crates/block-producer/src/tx_pool/mod.rs

```diff
@@ -0,0 +1,514 @@
+#![allow(unused)]
+
+use std::{
+    collections::{btree_map::Entry, BTreeMap, BTreeSet},
+    fmt::Display,
+    sync::Arc,
+};
+
+use miden_objects::{
+    accounts::AccountId,
+    transaction::{ProvenTransaction, TransactionId},
+    Digest,
+};
+
+use crate::store::TransactionInputs;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct BatchId(u64);
```

Comment:
For batch ID we can use a hash of all transaction IDs that go into the batch (using some stable order). This way, two batches with the same set of transactions would end up having the same ID.

Reply:
This would be ideal but I have concerns over how expensive that hash would be? This has to be performed inline meaning it extends the transaction pool lock.

Reply:
I think this depends on which hash function we use:

- If we go with something like BLAKE3, computing a hash of 16 transaction IDs should take less than 1 microsecond.
- If we use our arithmetization-friendly hash (RPO) it would be around 50 microseconds (or around 25 microseconds if/when we migrate to RPX hash function).

We could probably start with BLAKE3 and switch to RPO/RPX only if needed, but I think either way it may be OK as my expectation was that batch selection will probably take on the order of 1 millisecond anyways.

Reply:
I've left this as is still. I think we may want to separate the concept of batch ID and batch job ID for the situation where we drop a batch and re-create the exact same one. This would result in two "inflight" batches with the same ID.

Maybe this is fine; will have to think about it some more. For now both count and hashed are still around.

Reply:
> I think we may want to separate the concept of batch ID and batch job ID for the situation where we drop a batch and re-create the exact same one. This would result in two "inflight" batches with the same ID.

As far as I can tell, this shouldn't happen because we can't put the same transaction into two different batches at the same time (or at least we shouldn't). So, we could have two batches with the same ID - but it shouldn't be at the same time.

My preference is still to use a hash-based ID as I think it is conceptually a bit simpler and could help us catch some errors (i.e., if we do detect 2 batches with the same ID in flight, something probably went wrong).

Reply:
An example:

Two inflight batches `N` and `N+1` where `N+1` depends on `N`. 

Batch `N` fails to prove and we inform the mempool of this. The mempool requeues all transactions from `N` and `N+1`.

Select two new batches and we get `N` and `N+1` again. We now have three batches inflight, with a duplicate `N+1`.

Note that the batch producer has no idea of any dependencies so it cannot automatically drop the original `N+1`.